piton/tools/bin/riscvlib.py

```diff
@@ -0,0 +1,210 @@
+# Copyright and related rights are licensed under the Solderpad Hardware
+# License, Version 0.51 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+# http://solderpad.org/licenses/SHL-0.51. Unless required by applicable law
+# or agreed to in writing, software, hardware and materials distributed under
+# this License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+#
+# Author: Michael Schaffner <schaffner@iis.ee.ethz.ch>, ETH Zurich
+# Date: 04.02.2019
+# Description: Device tree generation script for OpenPiton+Ariane.
+
+
+import pyhplib
+from pyhplib import *
+
+import time
+
+# format reg leaf entry
+def regFmt(addrBase, addrLen, addrCells, sizeCells):         
+    
+    assert addrCells >= 1 or addrCells <= 2
+    assert sizeCells >= 0 or sizeCells <= 2
+
+    tmpStr = " "
+    
+    if addrCells >= 2:
+        tmpStr += "0x%08X " % (addrBase >> 32)
+    
+    if addrCells >= 1:
+        tmpStr += "0x%08X " % (addrBase & 0xFFFFFFFF)
+    
+    if sizeCells >= 2:
+        tmpStr += "0x%08X " % (addrLen >> 32)
+    
+    if sizeCells >= 1:
+        tmpStr += "0x%08X " % (addrLen & 0xFFFFFFFF)
+            
+    return tmpStr
+
+
+def gen_riscv_dts(devices, nCpus, cpuFreq, timeBaseFreq, periphFreq, dtsPath):
+
+    assert nCpus >= 1
+
+    tmpStr = '''// DTS generated with gen_riscv_dts(...) 
+// OpenPiton + Ariane framework
+// Date: %s
+
+/dts-v1/;
+
+/ {
+    #address-cells = <2>;
+    #size-cells = <2>;
+    compatible = "eth,ariane-bare-dev";
+    model = "eth,ariane-bare";
+    cpus {
+        #address-cells = <1>;
+        #size-cells = <0>;
+        timebase-frequency = <%d>; 
+    ''' % (time.strftime("%b %d %Y %H:%M:%S", time.localtime()), timeBaseFreq)
+
+    for k in range(nCpus):
+        tmpStr += '''    
+        CPU%d: cpu@%d {
+            clock-frequency = <%d>; 
+            device_type = "cpu";
+            reg = <%d>;
+            status = "okay";
+            compatible = "eth, ariane", "riscv";
+            riscv,isa = "rv64imacsu";
+            mmu-type = "riscv,sv39";
+            tlb-split;
+            // HLIC - hart local interrupt controller
+            CPU%d_intc: interrupt-controller {
+                #interrupt-cells = <1>;
+                interrupt-controller;
+                compatible = "riscv,cpu-intc";
+            };
+        };
+        ''' % (k,k,cpuFreq,k,k)
+
+    tmpStr += '''
+    };
+    '''
+
+    # this parses the device structure read from the OpenPiton devices*.xml file
+    # only get main memory ranges here
+    for i in range(len(devices)):
+        if devices[i]["name"] == "mem":
+            addrBase = devices[i]["base"]
+            addrLen  = devices[i]["length"]
+            tmpStr += '''
+    memory@%08X {
```

Comment:
dtc complains for me about this being `memory@%08X` because the output hex numbers are uppercase - can we change each of them to `@%08x` instead?

Reply:
ok I pushed a fix for this